pgwire: set options based on "options" URL parameter

mneverov opened the pull request:
pgwire: set options based on "options" URL parameter

Previously, CRDB ignored "options" URL parameter. User session parameters should
have been set via URL parameters directly:
`postgres://user@host:port/database?serial_normalization=virtual_sequence`

CRDB can now parse "options" URL parameter and set corresponding session
parameters (in compliance with Postgres jdbc connection parameters):
`postgres://user@host:port/database?options=-c%20serial_normalization=virtual_sequence`

Fixes #59404

Release note (sql change): CockroachDB now recognizes "options" URL parameter.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/59621)
<!-- Reviewable:end -->

mneverov commented:
To the reviewer:
To test the connection string I added a test that does nothing but tries to connect to the running CRDB server. The server response is not validated, only the session arguments. That is why I close the connection unconditionally right after the `Ping` is invoked. I also added some more test cases for `parseClientProvidedSessionParameters`.

Current implementation treats `user` and `crdb:remote_addr` as unknown configuration parameters (but not `results_buffer_size`) when configured in `options`. This is done according to the existing [session variables](https://github.com/cockroachdb/cockroach/blob/171cc86d6d84c110197485b5d886d77cddc12399/pkg/sql/vars.go#L122). Please, let me know if it should be changed.

Postgres does not allow `database` to be set up in `options` as it is not PGC_USERSET option. CRDB does have [Cockroach extension](https://github.com/cockroachdb/cockroach/blob/171cc86d6d84c110197485b5d886d77cddc12399/pkg/sql/vars.go#L207), that is database can be set via `options`. Let me know if `database` in options should be handled differently.

Lastly, as the connection string can contain multiple identical URL parameters (keys) with different values the current implementation to handle this is "last write win" (as it was before this PR).

mneverov commented:
I believe `TestCCLLogic` has nothing to do with my changes

rafiss commented:
Thanks for your contribution! I'll review this soon

rafiss reviewed (changes requested):
thank you again! i have requests for testing

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59621#-:-MS_HtbD3ZcN9U3_RXFC:b4dqul7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mneverov)

___
*[pkg/sql/pgwire/conn_test.go, line 524 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59621#-MS_DOo5B8oILaBOVde-:-MS_DOo63nRJ4M4A05C7:bvpa34c) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea6398d11675d1b3f6ef3cf39b22f5b5f84be139/pkg/sql/pgwire/conn_test.go#L524)):*
> ```Go
> // handshake. This emulates what pgwire.Server does.
> func waitForClientConn(ln net.Listener) (*conn, error) {
> 	conn, _, err := getSessionArgs(ln, false)
> ```

can you comment this like: `getSessionArgs(ln, false /* trustRemoteAddr */)`
___
*[pkg/sql/pgwire/conn_test.go, line 1266 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59621#-MS_79Ds2s9x6yAU7rx9:-MS_79Ds2s9x6yAU7rxA:b-qwi4bh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea6398d11675d1b3f6ef3cf39b22f5b5f84be139/pkg/sql/pgwire/conn_test.go#L1266)):*
> ```Go
> }
> 
> func TestParseClientProvidedSessionParameters(t *testing.T) {
> ```

this is an excellent unit test for `parseClientProvidedSessionParameters`

i think we should also have at least one end-to-end test. that is, we should connect using the string, then run a command like `SHOW serial_normalization` and makes sure the value specified in the string was set
___
*[pkg/sql/pgwire/conn_test.go, line 1294 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59621#-MS_6WgY49x3jFIg3F-n:-MS_6WgZ0M181u6dmacs:befohq0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea6398d11675d1b3f6ef3cf39b22f5b5f84be139/pkg/sql/pgwire/conn_test.go#L1294)):*
> ```Go
> 			query: "user=root",
> 			assert: func(t *testing.T, args sql.SessionArgs, err error) {
> 				assert.NoError(t, err)
> ```

we generally use the `require` package, as that would cause the test to abort on failure: https://godoc.org/github.com/stretchr/testify/require
___
*[pkg/sql/pgwire/conn_test.go, line 1337 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59621#-MS_Hh_97aOimJojREdo:-MS_Hh_97aOimJojREdp:b-uutynu) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea6398d11675d1b3f6ef3cf39b22f5b5f84be139/pkg/sql/pgwire/conn_test.go#L1337)):*
> ```Go
> 			assert: func(t *testing.T, args sql.SessionArgs, err error) {
> 				assert.Error(t, err)
> 				assert.Regexp(t, "is invalid, check '='", err)
> ```

should the regexp say `option "search_path" is invalid` ?
___
*[pkg/sql/pgwire/conn_test.go, line 1342 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59621#-MS_FmrL6oKjfc6eHFb0:-MS_FmrL6oKjfc6eHFb1:b-yxcru0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea6398d11675d1b3f6ef3cf39b22f5b5f84be139/pkg/sql/pgwire/conn_test.go#L1342)):*
> ```Go
> 		{
> 			desc:  "parse options successfully",
> 			query: "user=root&options=-c%20search_path=default,test%20-c%20optimizer_use_multicol_stats=true",
> ```

i realized that it also needs to work when the `=` sign is url-encoded.

could you add a test for: `options=-c%20search_path%3dtest`


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59621)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mneverov)

___
*[pkg/sql/pgwire/server.go, line 766 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59621#-MS_J79KDnN_e2xcPSxy:-MS_J79KDnN_e2xcPSxz:b-9gn5ka) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea6398d11675d1b3f6ef3cf39b22f5b5f84be139/pkg/sql/pgwire/server.go#L766)):*
> ```Go
> 			}
> 
> 			options := strings.Split(optionsRaw, "-c ")
> ```

i just learned that the space after `-c` is not required. e.g. `options=-csearch_path=test` should also work. there should be a test for this too


<!-- Sent from Reviewable.io -->

rafiss commented:
The release note could also use more information:

```
Release note (sql change): CockroachDB now recognizes "options" URL parameter. The
"options" parameter specifies session variables to set at connection start. This
is treated the same as defined in the PostgreSQL docs: https://www.postgresql.org/docs/13/libpq-connect.html#LIBPQ-PARAMKEYWORDS
```

mneverov commented:
Hi @rafiss , thanks for the review!
While addressing the comments I found that the following option formats are also fine:
`--search_path=public`
`-csearch-path=public` see [ParseLongOption](https://github.com/postgres/postgres/blob/2f86ab305e7fbc7b84960079551cf9cafd29684f/src/backend/utils/misc/guc.c#L10690)
`-d 1` see [process_postgres_switches](https://github.com/postgres/postgres/blob/2f86ab305e7fbc7b84960079551cf9cafd29684f/src/backend/tcop/postgres.c#L3654)

From the list of the session arguments available in CRDB and in PG I see the only common short option is `e` - for `datestyle`. Does it make sense to support short options for `datestyle`?
Does it make sense to support `--arg_name` style and convert `-` to `_`?
Also, I can't find better place for the parsing method than `server`, but it gets a bit more complex than now. Do you think it is ok to leave it there?

rafiss commented:
@mneverov wow, very interesting finds. I think for now let's keep this PR simple:
1. no need to support short options right now
2. no need to convert `-` to `_`
3. no need to handle `--arg_name` style. just `-c arg_name` is fine.

if you do think of a simple way to do (2) or (3) then go ahead. but it's not necessary since even the functionality as-is unblocks the PGJDBC use case that we were initially interested in solving. we can add them later if needed.

also, keeping it in server.go is fine, but can you create a new function in the same file for the logic to split the options? (the code in `case "options":`

mneverov reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59621)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/sql/pgwire/conn_test.go, line 524 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59621#-MS_DOo5B8oILaBOVde-:-MSabplX1NcpEG4b07zD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea6398d11675d1b3f6ef3cf39b22f5b5f84be139/pkg/sql/pgwire/conn_test.go#L524)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

can you comment this like: `getSessionArgs(ln, false /* trustRemoteAddr */)`
</blockquote></details>

Done.
___
*[pkg/sql/pgwire/conn_test.go, line 1266 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59621#-MS_79Ds2s9x6yAU7rx9:-MSrWdKFDvbv52Dz7-E-:blkhflv) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea6398d11675d1b3f6ef3cf39b22f5b5f84be139/pkg/sql/pgwire/conn_test.go#L1266)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this is an excellent unit test for `parseClientProvidedSessionParameters`

i think we should also have at least one end-to-end test. that is, we should connect using the string, then run a command like `SHOW serial_normalization` and makes sure the value specified in the string was set


</blockquote></details>

Added a separate e2e test for that.
___
*[pkg/sql/pgwire/conn_test.go, line 1294 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59621#-MS_6WgY49x3jFIg3F-n:-MSabwjS56dp4Kt3GLf6:b3vkw2p) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea6398d11675d1b3f6ef3cf39b22f5b5f84be139/pkg/sql/pgwire/conn_test.go#L1294)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

we generally use the `require` package, as that would cause the test to abort on failure: https://godoc.org/github.com/stretchr/testify/require
</blockquote></details>

fixed, thx
___
*[pkg/sql/pgwire/conn_test.go, line 1337 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59621#-MS_Hh_97aOimJojREdo:-MSagjzO4XqdUuhkZ6-W:bp9c7w2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea6398d11675d1b3f6ef3cf39b22f5b5f84be139/pkg/sql/pgwire/conn_test.go#L1337)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

should the regexp say `option "search_path" is invalid` ?
</blockquote></details>

Fixed.
The invalid option with the value is printed in this case. I checked the vars file, looks like there is no sensitive info can be exposed (no passwords) this way. 
___
*[pkg/sql/pgwire/conn_test.go, line 1342 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59621#-MS_FmrL6oKjfc6eHFb0:-MSaiYuIEsavRdtq7kE4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea6398d11675d1b3f6ef3cf39b22f5b5f84be139/pkg/sql/pgwire/conn_test.go#L1342)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i realized that it also needs to work when the `=` sign is url-encoded.

could you add a test for: `options=-c%20search_path%3dtest`
</blockquote></details>

Done.
___
*[pkg/sql/pgwire/server.go, line 766 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59621#-MS_J79KDnN_e2xcPSxy:-MSrWmV77oQEw3sr055s:b9r1ph1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea6398d11675d1b3f6ef3cf39b22f5b5f84be139/pkg/sql/pgwire/server.go#L766)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i just learned that the space after `-c` is not required. e.g. `options=-csearch_path=test` should also work. there should be a test for this too
</blockquote></details>

As discussed, added cases for `-ckey=val`, `-c key=val`, `--key=val`.


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
this looks great! almost ready to go, but just have one question and style comments.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59621#-:-MT2X7YjBVvlf_w6Iq4s:b-veov0m)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mneverov and @rafiss)

___
*[pkg/sql/pgwire/conn_test.go, line 1497 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59621#-MT2Wvzh0jjVlszq_YMb:-MT2Wvzh0jjVlszq_YMc:bjmfh26) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fae8c5bd96f2b9dbdf21d01c709406f12db1a5d/pkg/sql/pgwire/conn_test.go#L1497)):*
> ```Go
> 		// user is not a session variable; not an error, should log
> 		// "unknown configuration parameter"
> 		"user":        "",
> ```

i think `user` should not be in this map. why is it present?
___
*[pkg/sql/pgwire/server.go, line 764 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59621#-MT2WPSA5xEa912Q95np:-MT2WPSA5xEa912Q95nq:b6x0b4m) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fae8c5bd96f2b9dbdf21d01c709406f12db1a5d/pkg/sql/pgwire/server.go#L764)):*
> ```Go
> 				return sql.SessionArgs{}, err
> 			}
> 			for i := range opts {
> ```

small style suggestion:

```
for _, opt := range opts {
  err = loadParameter(ctx, opt.key, opt.value, &args)
  // ...
}
```
___
*[pkg/sql/pgwire/server.go, line 832 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59621#-MT2VTJWEpvKcsg4vHf-:-MT2VTJWEpvKcsg4vHf0:blym77k) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fae8c5bd96f2b9dbdf21d01c709406f12db1a5d/pkg/sql/pgwire/server.go#L832)):*
> ```Go
> 	opts := splitOptions(optionsRaw)
> 
> 	for i := 0; i < len(opts); i++ {
> ```

would this be more readable with a loop like `for _, opt := range opts {}`?
___
*[pkg/sql/pgwire/server.go, line 833 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59621#-MT2UL49EG_ofDrTlDUE:-MT2UL49EG_ofDrTlDUF:b-1xmcta) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fae8c5bd96f2b9dbdf21d01c709406f12db1a5d/pkg/sql/pgwire/server.go#L833)):*
> ```Go
> 
> 	for i := 0; i < len(opts); i++ {
> 		if opts[i] == "-c" {
> ```

minor style comment: could this be a `switch` statement to be a bit more readable?

maybe like this

```
for i := 0; i < len(opts); i++ {
  prefix := ""
  if len(opts[i]) > 1 {
    prefix = opts[i][:2]
  }

  switch {
  case opts[i] == "-c":
    // ...
  case lastWasDashC:
    // ...
  case prefix == "--" || prefix == "-c":
    // ...
  default:
    return nil, pgerror.Newf(...)
  }
}
return res, nil
```



<!-- Sent from Reviewable.io -->

mneverov committed:
pgwire: set options based on "options" URL parameter

Previously, CRDB ignored "options" URL parameter. User session parameters should
have been set via URL parameters directly:
`postgres://user@host:port/database?serial_normalization=virtual_sequence`

CRDB can now parse "options" URL parameter and set corresponding session
parameters (in compliance with Postgres jdbc connection parameters):
`postgres://user@host:port/database?options=-c%20serial_normalization=virtual_sequence`

Fixes #59404

Release note (sql change): CockroachDB now recognizes "options" URL parameter. The
"options" parameter specifies session variables to set at connection start. This
is treated the same as defined in the PostgreSQL docs: https://www.postgresql.org/docs/13/libpq-connect.html#LIBPQ-PARAMKEYWORDS

mneverov reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59621)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/sql/pgwire/conn_test.go, line 1497 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59621#-MT2Wvzh0jjVlszq_YMb:-MT6sIdq3Xbn3FoWeUqT:b-5z68og) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fae8c5bd96f2b9dbdf21d01c709406f12db1a5d/pkg/sql/pgwire/conn_test.go#L1497)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think `user` should not be in this map. why is it present?
</blockquote></details>

I thought adding a comment that "user" is not a session variable sets up expectations, but you're right, it's more confusing than helpful. Removed.
___
*[pkg/sql/pgwire/server.go, line 764 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59621#-MT2WPSA5xEa912Q95np:-MT6tWaN-jsO9SSPnsby:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fae8c5bd96f2b9dbdf21d01c709406f12db1a5d/pkg/sql/pgwire/server.go#L764)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

small style suggestion:

```
for _, opt := range opts {
  err = loadParameter(ctx, opt.key, opt.value, &args)
  // ...
}
```
</blockquote></details>

Done.
___
*[pkg/sql/pgwire/server.go, line 832 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59621#-MT2VTJWEpvKcsg4vHf-:-MT7-36p54PtK772wZlP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fae8c5bd96f2b9dbdf21d01c709406f12db1a5d/pkg/sql/pgwire/server.go#L832)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

would this be more readable with a loop like `for _, opt := range opts {}`?
</blockquote></details>

Done.
___
*[pkg/sql/pgwire/server.go, line 833 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59621#-MT2UL49EG_ofDrTlDUE:-MT7-5zs5eMIgETeW7wc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fae8c5bd96f2b9dbdf21d01c709406f12db1a5d/pkg/sql/pgwire/server.go#L833)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

minor style comment: could this be a `switch` statement to be a bit more readable?

maybe like this

```
for i := 0; i < len(opts); i++ {
  prefix := ""
  if len(opts[i]) > 1 {
    prefix = opts[i][:2]
  }

  switch {
  case opts[i] == "-c":
    // ...
  case lastWasDashC:
    // ...
  case prefix == "--" || prefix == "-c":
    // ...
  default:
    return nil, pgerror.Newf(...)
  }
}
return res, nil
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
this looks great. thanks again for your contribution! feel free to message me in the community slack if you'd like to receive a gift from us to show our appreciation

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59621#-:-MT70gls2A6auWs-fggK:b-wy85u0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)



<!-- Sent from Reviewable.io -->
